projecttacoma/fqm-execution
# Cache ElmJsons

# Summary
Added optional caching for cqls, elmJSONs, rootLibIdentifier, codeService, rep, vsMap when running execution

## New behavior
When passed the `useElmJsonsCaching` option, execution will check the existing ELMInfoCache to see if a measure with the same id and version has already been run in the last 10 minutes. It will use the previously extracted results for `cqls, elmJSONs, rootLibIdentifier, codeService, rep, vsMap` during the execution if the results exist and are valid.

## Code changes
- New file `ElmInfoCache` storing the cache, and functions to retrieve info from it as well as clear it
- Added `useElmJsonsCaching` option on calculation options to enable caching
- Added 'clearElmJsonsCache` option to force clear the existing ElmInfoCache before calculation
- New cli option `cache-elm` to enable ElmJsons caching
- Added testing

# Testing guidance
- `npm run check`
- Run a measure report calculation with the `cache-elm` option set to true and ensure that behavior is unchanged
- From `fqm-execution`, run `npm link`
- Open fqm-testify and run `rm -rf node_modules/fqm-execution` then `npm link fqm-execution`
- In fqm-testify in `util/MeasureCalculation.ts`, add `useElmJsonsCaching: true` following to the `options` const on line 15
- `npm run dev`
- Upload a measure
- Create a patient, select and ensure calculation results are correct
- Create a test resource and ensure calculation results are correct
- Use the chrome profiler to see the reduction in calculation time when caching is enabled. If you are unsure how to do this, reach out to me and I can walk you through!
